### PR TITLE
Node local partition pruning

### PR DESCRIPTION
**Pull request by lxynov**
Create empty page source for splits unmatching dynamic filter

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java [hunk 1]
```diff
@@ -301,10 +325,10 @@ public ColumnMappingKind getKind()
             return kind;
         }
 
-        public String getPrefilledValue()
+        public Object getPrefilledValue()
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: could you return `Optional<Object>` instead?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java [hunk 2]
```diff
@@ -301,10 +325,10 @@ public ColumnMappingKind getKind()
             return kind;
         }
 
-        public String getPrefilledValue()
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: this should be marked as `@Nullable`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java [hunk 3]
```diff
@@ -160,60 +131,7 @@ public HivePageSource(
             }
 
             if (columnMapping.getKind() == PREFILLED) {
-                String columnValue = columnMapping.getPrefilledValue();
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java]: extract this to separate commit, e.g:
```
Reuse prefilled value in HivePageSource`
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java [hunk 4]
```diff
@@ -115,7 +117,8 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
                 hiveSplit.getFileSize(),
                 hiveSplit.getFileModifiedTime(),
                 hiveSplit.getSchema(),
-                hiveTable.getCompactEffectivePredicate().intersect(dynamicFilter.transform(HiveColumnHandle.class::cast)),
+                hiveTable.getCompactEffectivePredicate(),
+                dynamicFilter.transform(HiveColumnHandle.class::cast),
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: why this split is needed? IMO intersected domain will be even smaller than `dynamicFilter` on its own

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java [hunk 5]
```diff
@@ -162,7 +166,27 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
...
                 bucketNumber,
                 fileSize,
-                fileModifiedTime);
+                fileModifiedTime,
+                hiveStorageTimeZone);
+
+        // Check if prefilled values match dynamic filter
+        Map<HiveColumnHandle, Domain> dynamicFilterDomains = dynamicFilter.getDomains().get();
+        if (!dynamicFilterDomains.isEmpty()) {
+            List<ColumnMapping> prefilledColumnMappings = ColumnMapping.extractPrefilledColumnMappings(columnMappings);
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: this performs partition pruning on node using `dynamicFilter`. I think it's fine, but it will also be addressed by cluster wide dynamic filtering that is in progress.

I would make the mechanism more generic though, so that it works on intersection of `dynamicFilter` and `effectivePredicate`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java [hunk 6]
```diff
@@ -162,7 +166,27 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
                 path,
                 bucketNumber,
                 fileSize,
-                fileModifiedTime);
+                fileModifiedTime,
+                hiveStorageTimeZone);
+
+        // Check if prefilled values match dynamic filter
+        Map<HiveColumnHandle, Domain> dynamicFilterDomains = dynamicFilter.getDomains().get();
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: add tests for this change

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java (hunk 5 above)
Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: So I think if the cluster-wide dynamic filtering will be checked in soon, we don't need to check in this PR. Because it brings no benefits but overhead. I'd like to take time learning the current status of cluster-wide dynamic filtering work before I go ahead addressing this PR's comments. Please LMK if you have other advice.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java (hunk 4 above)
Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: I think we only need to check partition keys against `dynamicFilter`. Checking them against `effectivePredicate` may not be necessary.

Consider a case when `dynamicFilter` is `TupleDomain.all()` whereas `effectivePredicate` isn't. By separating them, the checking work is totally avoided (because `dynamicFilter` is `TupleDomain.all()`)

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java (hunk 5 above)
Comment (raunaqmorarka) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: I am actively working on https://github.com/prestosql/presto/pull/1072/ 
While the feature itself works, I'm currently working on addressing comments about the way in which dynamic filters are passed to the coordinator. I'm hoping to finish this soon.

Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java]: Get it. Thank you for the update! @raunaqmorarka

**Comment by lxynov** [comment]
@sopel39 Thanks for the review! I've added a test for this change.

**Comment by tooptoop4** [comment]
@lxynov  have you got a benchmark result on before/after query time?

**Comment by sopel39** [comment]
@lxynov would you work on this?

**Comment by sopel39** [comment]
@lxynov could you rebase? Would you continue on this PR?

**Comment by lxynov** [comment]
@sopel39 Sorry for the delay. I took another look and think adding tests can be non-trivial work. And I may not have bandwidth to work on this soon. Please feel free to take it up on your side if you'd like. Otherwise I have to postpone this to late November

**Comment by raunaqmorarka** [comment]
Superseded by https://github.com/trinodb/trino/pull/9869
